src/lem/eval.rs

```diff
@@ -1123,6 +1594,29 @@ fn apply_cont() -> Func {
                         let (saved_env, continuation, _foo, _foo) = decons4(cont);
                         return (result, saved_env, continuation, makethunk)
                     }
+                    Cont::Cproc => {
+                        let (cproc_name, unevaled_args, evaluated_args, cont) = decons4(cont);
+                        // accumulate the evaluated arg (`result`)
+                        let evaluated_args: Expr::Cons = cons2(result, evaluated_args);
+                        match unevaled_args.tag {
+                            Expr::Nil => {
+                                // nothing else to evaluate
+                                let (expr, env, cont, ctrl) = choose_cproc_call(cproc_name, evaluated_args, env, cont);
+                                return (expr, env, cont, ctrl);
+                            }
+                            Expr::Cons => {
+                                // pop the next argument that needs to be evaluated
+                                let (arg, unevaled_args) = decons2(unevaled_args);
+                                let cont: Cont::Cproc = cons4(
+                                    cproc_name,
+                                    unevaled_args,
+                                    evaluated_args,
+                                    cont
+                                );
+                                return (arg, env, cont, ret);
+                            }
+                        }
+                    }
```

Comment:
Apart from the fact that this is obviously a new `Cont` variant, what can you tell me about the accumulation of arguments, and how it differs from any other function application? 

I'm trying to see if there is some insight about sharing, specifically, the logic that gathers evaluated arguments before application. Could something shared be more efficient or would this amount to the same thing? 

@gabriel-barrett might have a comment as well.

Reply:
Although similar, it's different from the accumulation that will happen once we refactor function applications.
The accumulation that happens for function applications needs to bind a variable to the reduced argument. What we accumulate is not just the evaluated argument, but a binding.

For coprocessors it's different (simpler, cheaper). We just need to cons a new reduced argument to the cons list of already evaluated arguments.

Reply:
Fair. In some ways I wonder if there's a link between this and creating sub-environments (à la `letrec`, which relates your point about merging that with `let`) where we could populate all bindings for a function *at once*.

/cc @gabriel-barrett

Reply:
> we could populate all bindings for a function at once

That, alone, would need to be discussed in more detail. But I think my motivation for the `let`/`letrec` comes from a higher level of abstraction, reaching the UX. Basically, if `letrec` can achieve everything `let` does with (apparently) the same number of iterations, then I don't see why we should support both if that makes our circuit bigger